File: shell_sort/shell_sort.py
def shell_sort(array):
    """
    Sorts the given array of integers using the Shell Sort algorithm
    Time Complexity : O((len(array))^2)
    Space Complexity : O(len(array))
    :param array: A List of integers.
    :return: returns the array sorted
    """

    # calculate the gap using Knuth's formula
    gap = 1
    while gap < len(array) // 3:
        gap = (gap * 3) + 1

    while gap > 0:
        # using this gap, exchange elements while you can
        for idx in range(gap, len(array)):
            val_to_insert = array[idx]
            candidate_idx = idx

            # shift all bigger elements to the right, creating a hole
            while candidate_idx > gap - 1 and array[candidate_idx - gap] > val_to_insert:
                array[candidate_idx] = array[candidate_idx - gap]
                candidate_idx -= gap

            # insert our element at the hole
            array[candidate_idx] = val_to_insert

        # decrease gap, math alert
        gap = (gap - 1) // 3

    return array


def main():
    sample_arr = [1, -312, 4, 12, 3, 17, 2542, 20, 18]
    print(shell_sort(sample_arr))


if __name__ == '__main__':
    main()

File: shell_sort/ShellSort.cs
using System;

public class ShellSort
{
    public static void Main()
    {
        int[] data = new int[] {1000, 45, -45, 121, 47, 45, 65, 121, -1, 103, 45, 34};
        Console.WriteLine("Data to be sorted:");
        Print(data);
        Sort(data);
        Console.WriteLine("Sorted data:");
        Print(data);
    }

    public static void Sort(int[] data)
    {
        for (int i = data.Length / 2; i > 0; i /= 2)
        {
            for (int j = i; j < data.Length; ++j)
            {
                for (int k = j - i; k >= 0; k -= i)
                {
                    if (data[k+i] >= data[k])
                    {
                        break;
                    }
                    else
                    {
                        int temp = data[k];
                        data[k] = data[k+i];
                        data[k+i] = temp;
                    }
                }
            }
        }
    }

    public static void Print(int[] data)
    {
        foreach(int elem in data)
        {
            Console.Write(elem + " ");
        }
        Console.WriteLine("");
    }
}
